### PR TITLE
support '#' in PackageUrl.url

### PR DESCRIPTION
**Pull request by mingwli**
[DevOps#62243](https://dev.azure.com/ceapex/Engineering/_workitems/edit/62243)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5121)

### CODE REVIEW

File: src/docfx/restore/RestoreGitMap.cs [hunk 1]
```diff
@@ -67,7 +67,7 @@ private RestoreGitMap(DependencyLockProvider dependencyLockProvider, string docs
                     return (path, gitLock.Commit);
 
                 default:
-                    throw new NotSupportedException($"Unknown package url: '{packageUrl}'");
+                    throw new NotSupportedException($"Unknown package url: '{packageUrl.Url}', type: {packageUrl.Type}");
```
Comment (yufeih) [inline src/docfx/restore/RestoreGitMap.cs]: consider modify `PackageUrl.ToString` method

File: docs/specs/config.yml [hunk 2]
```diff
@@ -585,3 +585,23 @@ outputs:
...
     {"document_id": "780484e9-a367-97ab-ba7d-a19f3b378a80", "document_version_independent_id": "e9b3f4e2-1a78-4b5d-fbb4-03e0a33e38f5",}
+---
+# dependencyPackageUrl's branch defined in url has higher priority
+repos:
+  https://docs.com/branch-priority/1:
+    - files:
+        docfx.yml: |
+          dependencies:
+            crr:
+              url: https://docs.com/branch-priority/2#master
```
Comment (yufeih) [inline docs/specs/config.yml]: Do we want to throw here for expanded form? as the `url` here means repository url.

Comment (mingwli) [reply docs/specs/config.yml]: Still it feels that:
```json
{
    "url": "https://github.com/owner/repo#branch",
    "includeInBuild": true
}
```
comes more convenient

File: src/docfx/lib/path/PackageUrl.cs [hunk 3]
```diff
@@ -77,9 +72,25 @@ private static (string remote, string refspec) SplitGitUrl(string remoteUrl)
...
+            {
+                Type = PackageType.Git;
+                var (url, branch) = SplitGitUrl(Url);
+                Url = url;
+                Branch = branch ?? Branch;
+            }
+            else
+            {
+                Type = PackageType.Folder;
+                Path = Url;
```
Comment (OsmondJiang) [inline src/docfx/lib/path/PackageUrl.cs]: it has problem when we just use `{Path: xxx}`

Comment (OsmondJiang) [reply src/docfx/lib/path/PackageUrl.cs]: add a test for that

File: src/docfx/lib/path/PackageUrl.cs [hunk 4]
```diff
@@ -77,9 +72,25 @@ private static (string remote, string refspec) SplitGitUrl(string remoteUrl)
...
+
+        [OnDeserialized]
+        private void OnDeserialized(StreamingContext context)
+        {
+            if (UrlUtility.IsHttp(Url))
+            {
+                Type = PackageType.Git;
+                var (url, branch) = SplitGitUrl(Url);
+                Url = url;
+                Branch = branch ?? Branch;
```
Comment (OsmondJiang) [inline src/docfx/lib/path/PackageUrl.cs]: it's a little inconsistent, the user specified `Branch` should has higher priority than branch in url?

File: src/docfx/lib/path/PackageUrl.cs [hunk 5]
```diff
@@ -31,16 +34,8 @@ public PackageUrl()
...
-                Type = PackageType.Git;
-                (Url, Branch) = SplitGitUrl(url);
-            }
-            else
-            {
-                Type = PackageType.Folder;
-                Path = url;
-            }
+            Url = url;
+            OnDeserialized(s_streamingContext);
```
Comment (yufeih) [inline src/docfx/lib/path/PackageUrl.cs]: if populate calls OnDeserialized, is it still necessary to call here?

File: test/docfx.Test/restore/RestoreTest.cs [hunk 6]
```diff
@@ -4,33 +4,42 @@
...
+        [InlineData("https://visualstudio.com/dotnet/docfx", PackageType.Git, "https://visualstudio.com/dotnet/docfx", "master", null)]
+        [InlineData("https://github.com/dotnet/docfx#master", PackageType.Git, "https://github.com/dotnet/docfx", "master", null)]
+        [InlineData("https://github.com/dotnet/docfx#live", PackageType.Git, "https://github.com/dotnet/docfx", "live", null)]
+        [InlineData("https://github.com/dotnet/docfx#", PackageType.Git, "https://github.com/dotnet/docfx", "master", null)]
+        [InlineData("https://github.com/dotnet/docfx#986127a", PackageType.Git, "https://github.com/dotnet/docfx", "986127a", null)]
+        [InlineData("https://github.com/dotnet/docfx#a#a", PackageType.Git, "https://github.com/dotnet/docfx", "a#a", null)]
+        [InlineData("https://github.com/dotnet/docfx#a\\b/d<e>f*h|i%3C", PackageType.Git, "https://github.com/dotnet/docfx", "a\\b/d<e>f*h|i%3C", null)]
+        [InlineData("crr/local-path", PackageType.Folder, "crr/local-path", "master", "crr/local-path")]
+        public static void PackageUrlTest(
+            string remote,
```
Comment (yufeih) [inline test/docfx.Test/restore/RestoreTest.cs]: consider test json deserialize in this method

File: src/docfx/lib/path/PackageUrl.cs (hunk 5 above)
Comment (mingwli) [reply src/docfx/lib/path/PackageUrl.cs]: This one is called by  
image
which share the same logic with OnDeserialized. And OnDeserialized needs to be manually invoked here.

Comment (yufeih) [reply src/docfx/lib/path/PackageUrl.cs]: I see, streaming context could then be null

Comment (mingwli) [reply src/docfx/lib/path/PackageUrl.cs]: I would very much like to. But it's a `struct` 😂.

Comment (mingwli) [reply src/docfx/lib/path/PackageUrl.cs]: `StreamingContext.Context` is always null in both case

**Pull request merged**